edit(fast-data): update Fast Data architecture description (#468)
mia-platform/documentation: docs/fast_data/architecture.md

## Services

> ### Real-Time Updater (RTU)

Reviewer: ```suggestion
### Real Time Updater (RTU)
```

> The Real-Time Updater is responsible for listening to the events concerning changes to the SoR and updating the Projections accordingly. Optionally, the RTU can perform other activities:

Reviewer: ```suggestion
The Real Time Updater is responsible for listening to the events concerning changes to the SoR and updating the Projections accordingly. Optionally, the RTU can perform other activities:
```

### MongoDB

MongoDB is a non-relational data store, used for storing Projections, Single Views, and some intermediate data, namely `Projection Change` and `Single View Error` collections.
In the case of Fast Data, the Projection, and Single View collections are saved in a format that is compatible with the CRUD service, which opens a ton of possibilities for exposing new services and consuming data in innovative ways.

MongoDB is great for horizontal scalability.

#### CRUD Service

Since the MongoDB collections are compatible with the CRUD Service and are automatically registered by the Console, they are ready to be exposed as APIs and are easily customizable (e.g. adding indexes just requires a few clicks in the Console). This allows you to easily connect them to the CMS, your applications, analytics, and third-party software.

## Putting it all together

The Fast Data architecture can be configured in various ways depending on your needs. There are no hard constraints, but generally, all the services belong to the same Kubernetes namespace, while the MongoDB and Kafka instances can be on managed hosts, on the same cluster, on-premise, etc.

> The SoRs usually belong to a different network portion, far from all the processing logic, since they must be kept independent of all the Fast Data flow, with the only exception of the CDC being connected to them. 

Reviewer 2: Do we define somewhere what is a SoR? (System Of Records).  
Being the first time we mention it, should we add `(System Of Records)` after the acronym?

### Standard Architecture

The standard architecture is rather streamlined, with just a couple of pivot points. A regular flow of information follows this path:
1. The CDC emits an event stating that some data in the SoR has changed;

> 2. The RTU performs the normalization of the messages received by the CDC to select the ones of interest and makes them adhere to a standard of interest, and then stores the Projections on MongoDB;

Reviewer 2: ```suggestion
2. The RTU performs the normalization of the messages received by the CDC to select the ones of interest and makes them adhere to a standard of interest, and then stores the transformed events as Projections Records on MongoDB;
```
To distinguish that a projection means collection on DB, while projection record means a single data/row/document of such collection.

3. The RTU computes and emits a `Projection Change` and saves it on MongoDB;
4. The SVC reads the `Projection Change`by polling MongoDB. Then, it aggregates the Single View using the new data and stores it in MongoDB.

>    1. The SVC emits a `Single View Event` if it is configured to do so. 

Reviewer 2: ```suggestion
   1. The SVC emits a `Single View Update Event` if it is configured to do so. 
```

![Fast Data Standard Architecture](img/fastdata-architecture-standard.svg)

### Standard Architecture with an SV-Patch

:::info
Everything related to the Standard Architecture is also valid with an SV-Patch.
:::

> Aggregation is not the only way possible to update Single Views, there is also an alternative called SV-Patch. This kind of operation is strongly recommended when a field of a Projection that is in common with a vast portion of Single Views, is updated. With this operation, the Single View Creator performs a Mongo update starting from the update of a single Projection, without regenerating the whole Single View. For an SV-Patch, the flow of information is as follows:

Reviewer 2: ```suggestion
Aggregation is not the only way possible to update Single Views, there is also an alternative called SV-Patch. This kind of operation is strongly recommended when a field of a Projection record that is in common with a vast portion of Single Views, is updated. Example of this type of fields can be a zip code or an internal product code.  
With this operation, the Single View Creator performs a Mongo update starting from the update of a single Projection record, without regenerating the whole Single View. For an SV-Patch, the flow of information is as follows:
```

1. The CDC emits an event stating that some data in the SoR has changed;
2. The RTU performs the normalization of the messages received by the CDC to select the ones of interest and makes them adhere to a standard of interest, and then stores the Projections on MongoDB;
  1. The RTU emits a `Projection Update` event (only for lookup/ constants);

> 4. The RTU computes and emits a `Projection Change`  (only for non-lookup/ constants)and saves it on MongoDB;

Reviewer 2: ```suggestion
4. The RTU computes and emits a `Projection Change`  (only for non-lookup/ constants) and saves it on MongoDB;
```

> 5. A second SVC (for SV-Patch operations) consumes the `Projection Update` messages from the RTU then uses them to aggregate the Single View and stores it in MongoDB.

Reviewer 2: ```suggestion
5. A second SVC (dedicated to SV-Patch operations) that consumes the `Projection Update` messages from the RTU then uses them to aggregate the Single View and stores it in MongoDB.
```
What do you think?

   1. The SVC emits a `Single View Event` if it is configured to do so. 

![Fast Data Standard Architecture with an SV-Patch](img/fastdata-architecture-standard-svpatch.svg)

> Click on this [link](https://docs.mia-platform.eu/docs/fast_data/configuration/single_views#single-view-patch) for more details on SV-Patch configurations.

Reviewer: ```suggestion
Click on this [link](https://docs.mia-platform.eu/docs/fast_data/configuration/single_views.md#single-view-patch) for more details on SV-Patch configurations.
```

> Click on this [link](https://docs.mia-platform.eu/docs/fast_data/configuration/single_views#single-view-patch) for more details on SV-Patch configurations.

Reviewer: ```suggestion
Click on [this link](/fast_data/configuration/single_views.md#single-view-patch) for more details on SV-Patch configurations.
```

### Event-Driven Architecture

:::caution

> The Single View Trigger Generator (SVTG) is still a BETA Plugin and is under active development. Pay attention to using the following Event-Driven Architecture.

Reviewer 2: @TheRusCompass Are we going to maintain this banner or should we start removing it?

![Fast Data Event-Driven Architecture with RTU](img/fastdata-architecture-event-driven.svg)

### Alternative Event-Driven Architecture

:::caution
The Single View Trigger Generator (SVTG) is still a BETA Plugin and is under active development. Pay attention to using the following Alternative Event-Driven Architecture.
:::

For the sake of being retro-compatible, you can configure the SVTG to compute and emit `Projection Change` events instead of `sv-trigger` events, and save it on MongoDB. Then the SVC will use that `Projection Change` message to aggregate the Single View. Such an architecture will look like this:

![Fast Data Event-Driven Architecture with RTU - Retrocompatible](img/fastdata-architecture-alternative-event-driven.svg)

### Architecture with Bucket Storage Support

Bucket Storage Support can be seamlessly integrated into any Fast Data architecture by attaching it to the same projection ingestion topics coming from the CDC. Paired with a Standard Architecture, the flow of information will look similar to this:

![Fast Data Architecture with Bucket Storage Support](img/fastdata-architecture-bucketstoragesupport.svg)

> Click on [this link](/fast_data/bucket_storage_support/overview_bucket_storage_support.md) for more details on Bucket Storage Support.

Reviewer: ```suggestion
Click on [this link](/fast_data/bucket_storage_support/overview.md) for more details on Bucket Storage Support.
```

> Click on this [link](https://docs.mia-platform.eu/docs/fast_data/bucket_storage_support/overview_bucket_storage_support) for more details on Bucket Storage Support.

Reviewer: ```suggestion
Click on this [link](https://docs.mia-platform.eu/docs/fast_data/bucket_storage_support/overview_bucket_storage_support.md) for more details on Bucket Storage Support.
```

> Click on this [link](https://docs.mia-platform.eu/docs/fast_data/bucket_storage_support/overview_bucket_storage_support) for more details on Bucket Storage Support.

Reviewer: ```suggestion
Click on [this link](/fast_data/bucket_storage_support/overview_bucket_storage_support.md) for more details on Bucket Storage Support.
```
